### PR TITLE
Ports hitscan from TG, adds it to *SELECT* things

### PR DESCRIPTION
**Pull request by Qwertytoforty**
## What Does This PR Do

Ports hitscan from TG.
Ports get_line from tg.
AI sat turrets are no longer var edited.
The following things gets hitscan for now.
Pulse weapons, bar pulse shotgun ammo. Admin weapons, adds fear to DS.
AI sat turrets on LETHAL. Disabler / stun is unaffected. Another ai change included below.
Emitters, including emitters from malf AI, are included. Looks cool. Adds spice to poor malf AI
Death wand is hitscan, because it is funny, and limited in use.
Malf space drones from random event has it's immolator beams hitscan. A historically weak mob, seems the perfect target to test a mob with hitscan as a gimmick.

Other changes:
Malf AI turret upgrade now only upgrades the ai sat turrets, instead of all the turrets on the zlvl (lol)
Plasma pistol no longer shoots at people sometimes when discharging, instead always spews more plasma, avoids sleep issue.

AI turret changes:
Apart from hitscan, AI turrets also get a slight rework.
Due  to them being incredibly cheesable, they now have the ability to pierce 1 object / turf. They will not pierce mobs (for now)
This is to assist AI with locker cheese, and the fact turrets waste shots into corners on the sat so often.
As a trade off however, AI turret damage has been lowered to 7/8th of what it was before.
These lasers also won't hurt the AI, due to some positions you can stand in certain maps that would cause the laser to hit the AI through a wall.

## Why It's Good For The Game

Potentially interesting feature to carefully implement in some cases. 
As a rule of thumb:

> Things hitscan is good for: Stationary (turrets, emitter), broken by admin only (ds), rare in that one could be achived in a round and would have downsides (think a gimmicky anomaly core weapon), a simple mob with a weak hitscan as a gimmick (note they can be blocked)
things hitscan is not good for: common weapons, mobs unless a super low damage attack, remeber with hitscan, *mobs will not miss*.

Pulse weapons are good, they lack a certain fear for end of round vs biohazards, and are admin only. Emitters are good being stationary, and look cool.
AI turrets are good. Will help with the weakness of the turrets, especially with piercing.
Death wand is entirely because it's funny, and only 3 shots (6 ish with charge), and because I spent 2 days trying to figure out why it's colour was wrong.
Malf drone is due to it's low damage, and I want it to be a bit more of a threat, and it being a good test for hitscan mobs.

Plasma pistol changed due to sleep issue, as well most of the time it shot the person attacking the user, not much of a downside. This should make it hurt the user more, unless they are atmos proof.

## Images

image

image

image

## Testing

Lots of lag testing with reflectors. Lots of shooting skrells.

## Changelog
:cl:
add: Adds hitscan to: AI turrets, pulse lasers (bar shotgun), emitters, malf space drones, wand of death.
add: Ai sat turrets now pierce 1 non mob, and do slightly less damage.
fix: ai upload no longer upgraded by malf ai turret upgrade
tweak: plasma pistol no longer shoots other people on discharge instead always spews plasma, more plasma than before
/:cl:

### CODE REVIEW

**Comment by 1080pCat** [comment]
> Death wand is hitscan, because it is funny, and limited in use.

ungodly based, though this may add some skill requirements..


> add: Ai sat turrets now pierce 1 non mob, and do slightly less damage.

tweak would probably be better than add

**Comment by Adrer** [comment]
Images of changes would be appreciated so we can actually see it in action.

**Comment by AffectedArc07** [comment]
How do hitscan emitters behave with reflective blobs?

**Comment by Qwertytoforty** [comment]
> How do hitscan emitters behave with reflective blobs?

discussed on discord, behave the same as lasers, but have some extra precautions to avoid lag machines

File: code/game/objects/effects/temporary_visuals/projectile/projectile_effects.dm [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+/obj/effect/projectile/singularity_act()
+	return
+
+/obj/effect/projectile/proc/scale_to(nx,ny,override=TRUE)
+	var/matrix/M
+	if(!override)
+		M = transform
+	else
+		M = new
+	M.Scale(nx,ny)
```
Comment (Henri215) [inline code/game/objects/effects/temporary_visuals/projectile/projectile_effects.dm:21]:
```suggestion
	M.Scale(nx, ny)
```

File: code/game/objects/effects/temporary_visuals/projectile/projectile_effects.dm [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+		look.appearance_flags = RESET_COLOR | RESET_ALPHA
+	icon_state = null
+	add_overlay(look)
+	scale_to(1,scaling, FALSE)
+	turn_to(angle_override, FALSE)
+	if(!isnull(new_loc)) //If you want to null it just delete it...
+		forceMove(new_loc)
+	for(var/i in 1 to increment)
+		pixel_x += round((sin(angle_override)+16*sin(angle_override)*2), 1)
+		pixel_y += round((cos(angle_override)+16*cos(angle_override)*2), 1)
```
Comment (Henri215) [inline code/game/objects/effects/temporary_visuals/projectile/projectile_effects.dm:56]:
```suggestion
	scale_to(1, scaling, FALSE)
	turn_to(angle_override, FALSE)
	if(!isnull(new_loc)) //If you want to null it just delete it...
		forceMove(new_loc)
	for(var/i in 1 to increment)
		pixel_x += round((sin(angle_override) + 16 * sin(angle_override) * 2), 1)
		pixel_y += round((cos(angle_override) + 16 * cos(angle_override) * 2), 1)
```

**Comment by spacegeese** [comment]
"Fuck you!" death beam sounds really awesome

File: code/game/objects/effects/temporary_visuals/projectile/impact.dm [hunk 3]
```diff
@@ -0,0 +1,52 @@
+/obj/effect/projectile/impact
```
Comment (Contrabang) [inline code/game/objects/effects/temporary_visuals/projectile/impact.dm:1]: Can you add a comment above this explaining what impact projectiles are / how they are used as part of hitscan?

File: code/modules/projectiles/ammunition/energy_lens.dm [hunk 4]
```diff
@@ -42,14 +42,14 @@
...
-	projectile_type = /obj/item/projectile/beam/pulse
+	projectile_type = /obj/item/projectile/beam/pulse/hitscan
 	muzzle_flash_color = LIGHT_COLOR_DARKBLUE
 	e_cost = 200
 	select_name = "DESTROY"
 	fire_sound = 'sound/weapons/pulse.ogg'
 
 /obj/item/ammo_casing/energy/laser/scatter/pulse
-	projectile_type = /obj/item/projectile/beam/pulse
+	projectile_type = /obj/item/projectile/beam/pulse/hitscan
```
Comment (Contrabang) [inline code/modules/projectiles/ammunition/energy_lens.dm:52]: How does this work with scattering? Does it properly shoot out 5 hitscan beams in multiple directions? If it doesn't work, it should probably left as normal pulse rounds.

File: code/modules/projectiles/guns/energy/special_eguns.dm [hunk 5]
```diff
@@ -415,43 +415,12 @@
...
-	var/list/turf_targets = list()
-	for(var/turf/T in orange(get_turf(src), 7))
-		turf_targets += T
-	if(length(turf_targets))
-		var/turf/target = pick(turf_targets)
-		shootAt(target)
+	visible_message("<span class='danger'>[src] vents heated plasma!</span>")
+	var/turf/simulated/T = get_turf(src)
+	if(istype(T))
+		T.atmos_spawn_air(LINDA_SPAWN_HEAT | LINDA_SPAWN_TOXINS|LINDA_SPAWN_20C, 20)
```
Comment (Contrabang) [inline code/modules/projectiles/guns/energy/special_eguns.dm:421]: Can you mark down the changes to the plasma pistol in the changelog too?

File: code/modules/projectiles/projectile/beams.dm [hunk 6]
```diff
@@ -88,6 +140,22 @@
 	impact_effect_type = /obj/effect/temp_visual/impact_effect/green_laser
 	light_color = LIGHT_COLOR_GREEN
 
+/obj/item/projectile/beam/emitter/hitscan
```
Comment (Contrabang) [inline code/modules/projectiles/projectile/beams.dm:143]: This appears to be a duplicate of whats in special_projectiles.dm at line 286

File: code/modules/projectiles/projectile_base.dm [hunk 7]
```diff
@@ -1,3 +1,6 @@
+#define MOVES_HITSCAN -1 //Not actually hitscan but close as we get without actual hitscan.
+#define MUZZLE_EFFECT_PIXEL_INCREMENT 17 //How many pixels to move the muzzle flash up so your character doesn't look like they're shitting out lasers.
```
Comment (Contrabang) [inline code/modules/projectiles/projectile_base.dm:2]: Can you turn these into `///` comments and place them above the defines?

File: code/modules/projectiles/projectile_base.dm [hunk 8]
```diff
@@ -80,6 +84,32 @@
...
+	var/hitscan_light_intensity = 1.5
+	var/hitscan_light_range = 0.75
+	var/hitscan_light_color_override
+	var/muzzle_flash_intensity = 3
+	var/muzzle_flash_range = 1.5
+	var/muzzle_flash_color_override
+	var/impact_light_intensity = 3
+	var/impact_light_range = 2
+	var/impact_light_color_override
+	var/hitscan_duration = 0.3 SECONDS
```
Comment (Contrabang) [inline code/modules/projectiles/projectile_base.dm:111]: This is a lot of vars to add to every projectile, but I can't think of any better way besides creating something like /obj/item/projectile/hitscan. 

Also I'm just realizing, why the fuck are projectiles items?

File: code/modules/projectiles/projectile_base.dm [hunk 9]
```diff
@@ -400,11 +460,78 @@
...
 	return TRUE
+
+/obj/item/projectile/proc/record_hitscan_start(datum/point_precise/point_cache)
+	if(point_cache)
+		beam_segments = list()
+		beam_index = point_cache
+		beam_segments[beam_index] = null //record start.
+
+/obj/item/projectile/proc/process_hitscan()
+	var/safety = range * 10
```
Comment (Contrabang) [inline code/modules/projectiles/projectile_base.dm:483]: A comment here explaining exactly what safety represents here would be good

File: code/game/objects/effects/temporary_visuals/projectile/muzzle.dm [hunk 10]
```diff
@@ -0,0 +1,46 @@
+/obj/effect/projectile/muzzle
```
Comment (Contrabang) [inline code/game/objects/effects/temporary_visuals/projectile/muzzle.dm:1]: Ditto here

File: code/game/objects/effects/temporary_visuals/projectile/projectile_effects.dm [hunk 11]
```diff
@@ -0,0 +1,64 @@
+/obj/effect/projectile
```
Comment (Contrabang) [inline code/game/objects/effects/temporary_visuals/projectile/projectile_effects.dm:1]: ditto here too

File: code/game/objects/effects/temporary_visuals/projectile/tracer.dm [hunk 12]
```diff
@@ -0,0 +1,97 @@
+/proc/generate_tracer_between_points(datum/position/starting, datum/position/ending, beam_type, color, qdel_in = 5, light_range = 2, light_color_override, light_intensity = 1, instance_key) //Do not pass z-crossing points as that will not be properly (and likely will never be properly until it's absolutely needed) supported!
```
Comment (Contrabang) [inline code/game/objects/effects/temporary_visuals/projectile/tracer.dm:1]: and ditto above here

**Review by Contrabang — CHANGES_REQUESTED** [review]
Did my best reviewing the code here. A lot of the in-depth projectile code is confusing.

File: code/game/gamemodes/malfunction/Malf_Modules.dm [hunk 13]
```diff
@@ -329,11 +329,11 @@
 	unlock_sound = 'sound/items/rped.ogg'
 
 /datum/AI_Module/upgrade_turrets/upgrade(mob/living/silicon/ai/AI)
-	for(var/obj/machinery/porta_turret/turret in GLOB.machines)
+	for(var/obj/machinery/porta_turret/ai_turret/turret in GLOB.machines)
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/game/gamemodes/malfunction/Malf_Modules.dm:332]: I can, the description of the ability only mentioned it affecting the AI sat anyway

File: code/modules/projectiles/ammunition/energy_lens.dm (hunk 4 above)
Comment (Qwertytoforty) [reply code/modules/projectiles/ammunition/energy_lens.dm:52]: It works fine

File: code/__DEFINES/_math.dm [hunk 14]
```diff
@@ -244,3 +244,49 @@
 #define PERCENT_OF(val1, val2) (val1 * (val2 / 100))
 
 #define num2hex(X, len) uppertext(num2text(X, len, 16))
+
+/**
+ * Get a list of turfs in a line from `starting_atom` to `ending_atom`.
+ *
+ * Uses the ultra-fast [Bresenham Line-Drawing Algorithm](https://en.wikipedia.org/wiki/Bresenham%27s_line_algorithm).
+ */
+/proc/get_line(atom/starting_atom, atom/ending_atom)
```
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/__DEFINES/_math.dm:253]: Getline is only used 19 times, think it is worth converting it over then?

File: code/game/objects/effects/temporary_visuals/projectile/impact.dm (hunk 3 above)
Comment (Qwertytoforty) [reply code/game/objects/effects/temporary_visuals/projectile/impact.dm:1]: doen

File: code/modules/projectiles/projectile/beams.dm (hunk 6 above)
Comment (Qwertytoforty) [reply code/modules/projectiles/projectile/beams.dm:143]: duplicate removed

File: code/modules/projectiles/projectile_base.dm (hunk 8 above)
Comment (Qwertytoforty) [reply code/modules/projectiles/projectile_base.dm:111]: yeah

File: code/modules/projectiles/projectile_base.dm (hunk 9 above)
Comment (Qwertytoforty) [reply code/modules/projectiles/projectile_base.dm:483]: Done

File: code/game/objects/effects/temporary_visuals/projectile/muzzle.dm (hunk 10 above)
Comment (Qwertytoforty) [reply code/game/objects/effects/temporary_visuals/projectile/muzzle.dm:1]: done

File: code/game/objects/effects/temporary_visuals/projectile/projectile_effects.dm (hunk 11 above)
Comment (Qwertytoforty) [reply code/game/objects/effects/temporary_visuals/projectile/projectile_effects.dm:1]: done

File: code/game/objects/effects/temporary_visuals/projectile/tracer.dm (hunk 12 above)
Comment (Qwertytoforty) [reply code/game/objects/effects/temporary_visuals/projectile/tracer.dm:1]: done

File: code/modules/projectiles/projectile_base.dm (hunk 7 above)
Comment (Qwertytoforty) [reply code/modules/projectiles/projectile_base.dm:2]: done

**Comment by SpringSkipper** [comment]
should revival wand/healing staff also be hitscan to match the death wand?

**Review by Contrabang — APPROVED** [review]

File: code/__DEFINES/_math.dm (hunk 14 above)
[the remark this answers is not in the document]
Comment (Qwertytoforty) [reply code/__DEFINES/_math.dm]: was moved to a proper place

**Review by S34NW — APPROVED** [review]

**Pull request merged**